File: notas.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

type Nota struct {
	Titulo    string `json:"titulo"`
	Contenido string `json:"contenido"`
}

func NewNota(titulo, contenido string) *Nota {
	return &Nota{Titulo: titulo, Contenido: contenido}
}

type Categoria struct {
	Nombre string  `json:"nombre"`
	Notas  []*Nota `json:"notas"`
}

func NewCategoria(nombre string) *Categoria {
	return &Categoria{Nombre: nombre}
}

func (c *Categoria) AgregarNota(nota *Nota) {
	c.Notas = append(c.Notas, nota)
}

func (c *Categoria) MostrarNotas() {
	fmt.Printf("Notas en la categoría %s:\n", c.Nombre)
	for _, nota := range c.Notas {
		fmt.Printf("- %s\n", nota.Titulo)
	}
}

func (c *Categoria) BorrarNota(titulo string) {
	nuevasNotas := make([]*Nota, 0, len(c.Notas))
	for _, nota := range c.Notas {
		if nota.Titulo != titulo {
			nuevasNotas = append(nuevasNotas, nota)
		}
	}
	c.Notas = nuevasNotas
}

type NotasApp struct {
	Categorias map[string]*Categoria
}

func NewNotasApp() *NotasApp {
	app := &NotasApp{
		Categorias: make(map[string]*Categoria),
	}
	app.CargarNotasDesdeArchivo()
	return app
}

func (app *NotasApp) Bienvenida() {
	fmt.Println("¡Bienvenido a la aplicación de notas!")
}

func (app *NotasApp) NuevaNota() {
	titulo := leerEntrada("Ingrese el título de la nota: ")
	contenido := leerEntradaMultilinea("Ingrese el contenido de la nota (presione Enter para terminar):\n")
	categoria := leerEntrada("En qué categoría desea guardar la nota? ")

	if _, existe := app.Categorias[categoria]; !existe {
		crearCategoria := leerEntrada("La categoría no existe, ¿desea crearla? (s/n) ")
		if crearCategoria == "s" {
			app.Categorias[categoria] = NewCategoria(categoria)
		} else {
			fmt.Println("Nota no guardada.")
			return
		}
	}

	app.Categorias[categoria].AgregarNota(NewNota(titulo, contenido))
	app.GuardarNotasEnArchivo()
	fmt.Printf("Nota guardada exitosamente en la categoría %s\n", categoria)
}

func (app *NotasApp) BorrarNota() {
	fmt.Println("Categorías disponibles:")
	for nombre := range app.Categorias {
		fmt.Printf("- %s\n", nombre)
	}

	nombreCategoria := leerEntrada("Ingrese el nombre de la categoría de la nota que desea borrar: ")
	categoria, existe := app.Categorias[nombreCategoria]
	if !existe {
		fmt.Println("Categoría no encontrada.")
		return
	}

	categoria.MostrarNotas()
	tituloNota := leerEntrada("Ingrese el título de la nota que desea borrar: ")
	categoria.BorrarNota(tituloNota)
	app.GuardarNotasEnArchivo()
	fmt.Println("Nota borrada exitosamente.")
}

func (app *NotasApp) CargarNotasDesdeArchivo() {
	file, err := os.Open("notas.json")
	if err != nil {
		if !os.IsNotExist(err) {
			fmt.Println("Error al abrir el archivo de notas:", err)
		}
		return
	}
	defer file.Close()

	err = json.NewDecoder(file).Decode(&app.Categorias)
	if err != nil {
		fmt.Println("Error al decodificar el archivo de notas:", err)
	}
}

func (app *NotasApp) GuardarNotasEnArchivo() {
	file, err := os.Create("notas.json")
	if err != nil {
		fmt.Println("Error al crear el archivo de notas:", err)
		return
	}
	defer file.Close()

	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	err = encoder.Encode(app.Categorias)
	if err != nil {
		fmt.Println("Error al codificar el archivo de notas:", err)
	}
}

func leerEntrada(mensaje string) string {
	var entrada string
	fmt.Print(mensaje)
	fmt.Scanln(&entrada)
	return entrada
}

func leerEntradaMultilinea(mensaje string) string {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Print(mensaje)
	var lineas []string
	for scanner.Scan() {
		linea := scanner.Text()
		if linea == "" {
			break
		}
		lineas = append(lineas, linea)
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error al leer la entrada:", err)
	}
	return strings.Join(lineas, "\n")
}

func main() {
	app := NewNotasApp()
	app.Bienvenida()

	for {
		fmt.Println("\n¿Qué desea hacer?")
		fmt.Println("1. Crear una nueva nota")
		fmt.Println("2. Borrar una nota")
		fmt.Println("3. Salir del programa")
		opcion := leerEntrada("Ingrese el número de la opción: ")

		switch opcion {
		case "1":
			app.NuevaNota()
		case "2":
			app.BorrarNota()
		case "3":
			fmt.Println("¡Hasta luego!")
			return
		default:
			fmt.Println("Opción no válida. Intente de nuevo.")
		}
	}
}

File: notas.py
import json
import os

class Nota:
    def __init__(self, titulo, contenido):
        self.titulo = titulo
        self.contenido = contenido

    def to_dict(self):
        return {"titulo": self.titulo, "contenido": self.contenido}

    @staticmethod
    def from_dict(nota_dict):
        return Nota(nota_dict["titulo"], nota_dict["contenido"])

class Categoria:
    def __init__(self, nombre):
        self.nombre = nombre
        self.notas = []

    def agregar_nota(self, nota):
        self.notas.append(nota)

    def mostrar_notas(self):
        print(f"Notas en la categoría {self.nombre}:")
        for nota in self.notas:
            print(f"- {nota.titulo}")

    def borrar_nota(self, titulo):
        self.notas = [nota for nota in self.notas if nota.titulo != titulo]

    def to_dict(self):
        return {"nombre": self.nombre, "notas": [nota.to_dict() for nota in self.notas]}

    @staticmethod
    def from_dict(categoria_dict):
        categoria = Categoria(categoria_dict["nombre"])
        categoria.notas = [Nota.from_dict(nota_dict) for nota_dict in categoria_dict["notas"]]
        return categoria

class NotasApp:
    def __init__(self):
        self.categorias = {}
        self.cargar_notas_desde_archivo()

    def bienvenida(self):
        print("¡Bienvenido a la aplicación de notas!")

    def nueva_nota(self):
        titulo = input("Ingrese el título de la nota: ")
        contenido = input("Ingrese el contenido de la nota: ")

        categoria = input("En qué categoría desea guardar la nota? ")
        if categoria not in self.categorias:
            crear_categoria = input("La categoría no existe, ¿desea crearla? (s/n) ")
            if crear_categoria.lower() == "s":
                self.categorias[categoria] = Categoria(categoria)
            else:
                print("Nota no guardada.")
                return

        self.categorias[categoria].agregar_nota(Nota(titulo, contenido))
        self.guardar_notas_en_archivo()
        print(f"Nota guardada exitosamente en la categoría {categoria}")

    def borrar_nota(self):
        print("Categorías disponibles:")
        for nombre_categoria in self.categorias:
            print(f"- {nombre_categoria}")

        nombre_categoria = input("Ingrese el nombre de la categoría de la nota que desea borrar: ")
        if nombre_categoria not in self.categorias:
            print("Categoría no encontrada.")
            return

        categoria = self.categorias[nombre_categoria]
        categoria.mostrar_notas()
        titulo_nota = input("Ingrese el título de la nota que desea borrar: ")
        categoria.borrar_nota(titulo_nota)
        self.guardar_notas_en_archivo()
        print("Nota borrada exitosamente.")

    def cargar_notas_desde_archivo(self):
        if os.path.exists("notas.json"):
            with open("notas.json", "r") as file:
                data = json.load(file)
                for nombre_categoria, categoria_data in data.items():
                    self.categorias[nombre_categoria] = Categoria.from_dict(categoria_data)

    def guardar_notas_en_archivo(self):
        with open("notas.json", "w") as file:
            data = {nombre_categoria: categoria.to_dict() for nombre_categoria, categoria in self.categorias.items()}
            json.dump(data, file, indent=4)

    def iniciar(self):
        self.bienvenida()

        while True:
            print("\n¿Qué desea hacer?")
            print("1. Crear una nueva nota")
            print("2. Borrar una nota")
            print("3. Salir del programa")
            opcion = input("Ingrese el número de la opción: ")

            if opcion == "1":
                self.nueva_nota()
            elif opcion == "2":
                self.borrar_nota()
            elif opcion == "3":
                print("¡Hasta luego!")
                break
            else:
                print("Opción no válida. Intente de nuevo.")

if __name__ == "__main__":
    app = NotasApp()
    app.iniciar()
